File: packages/engram-ai-js/src/schema.ts
/**
 * Memory schema - matches Python engram-ai for cross-language compatibility.
 */

/**
 * Memory durability tier.
 */
export enum Durability {
  /** Stable facts that rarely change (name, preferences, biographical). */
  CORE = 'core',
  /** Temporary context with explicit end date (traveling, project work). */
  SITUATIONAL = 'situational',
  /** Things that happened, decays over time (conversations, events). */
  EPISODIC = 'episodic',
}

/**
 * Semantic type of memory content.
 */
export enum MemoryType {
  /** A factual statement (e.g., 'API rate limit is 100/min'). */
  FACT = 'fact',
  /** A rule or guideline to follow (e.g., 'Always use TypeScript strict mode'). */
  RULE = 'rule',
  /** A decision that was made (e.g., 'Chose Tailwind for utility-first CSS'). */
  DECISION = 'decision',
  /** A preference or like/dislike (e.g., 'Prefers dark mode'). */
  PREFERENCE = 'preference',
  /** Background context (e.g., 'Currently working on authentication refactor'). */
  CONTEXT = 'context',
  /** An observation or insight (e.g., 'User tends to ask follow-up questions'). */
  OBSERVATION = 'observation',
}

/**
 * How the memory was created.
 */
export enum MemorySource {
  /** User directly stated this fact. */
  EXPLICIT = 'explicit',
  /** Extracted/inferred from conversation. */
  INFERRED = 'inferred',
  /** Created by system (consolidation, migration, etc.). */
  SYSTEM = 'system',
}

/**
 * A memory item with full metadata.
 */
export interface Memory {
  /** Unique identifier for this memory. */
  id: string;
  /** The memory content (what we're remembering). */
  text: string;
  /** How permanent this memory is. */
  durability: Durability;
  /** Semantic type of this memory (fact, rule, decision, etc.). */
  memoryType?: MemoryType | null;
  /** Confidence score (0-1) for inferred memories. */
  confidence: number;
  /** How this memory was created. */
  source: MemorySource;
  /** When this fact became true. */
  validFrom: Date;
  /** When this fact expires (null = permanent). */
  validUntil?: Date | null;
  /** ID of the memory this one replaces (previous version). */
  supersedes?: string | null;
  /** ID of the memory that replaced this one (next version). */
  supersededBy?: string | null;
  /** When this memory was superseded. */
  supersededAt?: Date | null;
  /** When this memory was stored. */
  createdAt: Date;
  /** Last time this memory was retrieved (for decay calculations). */
  lastAccessedAt?: Date | null;
  /** Number of times this memory was retrieved. */
  accessCount: number;
  /** Optional tags for categorization. */
  tags: string[];
  /** Additional metadata. */
  metadata: Record<string, unknown>;
}

/**
 * Input for creating a new memory.
 */
export interface MemoryCreate {
  text: string;
  durability?: Durability;
  memoryType?: MemoryType | null;
  confidence?: number;
  source?: MemorySource;
  validFrom?: Date;
  validUntil?: Date | null;
  tags?: string[];
  metadata?: Record<string, unknown>;
}

/**
 * Input for updating a memory (creates new version in chain).
 */
export interface MemoryUpdate {
  text: string;
  confidence?: number;
  source?: MemorySource;
  validFrom?: Date;
  validUntil?: Date | null;
}

/**
 * Query parameters for memory retrieval.
 */
export interface MemoryQuery {
  /** Semantic search query. */
  query: string;
  /** Maximum number of results. */
  limit?: number;
  /** Minimum confidence threshold. */
  minConfidence?: number;
  /** Filter by durability tier(s). */
  durability?: Durability[];
  /** Filter by memory type(s). */
  memoryType?: MemoryType[];
  /** Include superseded (old version) memories. */
  includeSuperseded?: boolean;
  /** Include memories past their validUntil date. */
  includeExpired?: boolean;
  /** Filter by tags (any match). */
  tags?: string[];
  /** Check validity at this time (default: now). */
  validAt?: Date;
}

/**
 * Namespace tuple for scoping memories.
 */
export type Namespace = string[];

/**
 * Check if a memory is valid at a given time.
 */
export function isMemoryValid(memory: Memory, at: Date = new Date()): boolean {
  // Superseded memories are not valid
  if (memory.supersededBy != null) {
    return false;
  }

  // Check temporal validity
  if (memory.validFrom && at < memory.validFrom) {
    return false;
  }
  if (memory.validUntil && at > memory.validUntil) {
    return false;
  }

  return true;
}

/**
 * Check if this is the current version (not superseded).
 */
export function isMemoryCurrent(memory: Memory): boolean {
  return memory.supersededBy == null;
}

File: src/memable/schema.py
"""
Memory schema with durability tiers, temporal awareness, and version chains.
"""

from datetime import UTC, datetime
from enum import StrEnum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


def _utc_now() -> datetime:
    """Return current time as timezone-aware UTC datetime."""
    return datetime.now(UTC)


class Durability(StrEnum):
    """Memory durability tier."""

    CORE = "core"
    """Stable facts that rarely change (name, preferences, biographical)."""

    SITUATIONAL = "situational"
    """Temporary context with explicit end date (traveling, project work)."""

    EPISODIC = "episodic"
    """Things that happened, decays over time (conversations, events)."""


class MemoryType(StrEnum):
    """Semantic type of memory content."""

    FACT = "fact"
    """A factual statement (e.g., 'API rate limit is 100/min')."""

    RULE = "rule"
    """A rule or guideline to follow (e.g., 'Always use TypeScript strict mode')."""

    DECISION = "decision"
    """A decision that was made (e.g., 'Chose Tailwind for utility-first CSS')."""

    PREFERENCE = "preference"
    """A preference or like/dislike (e.g., 'Prefers dark mode')."""

    CONTEXT = "context"
    """Background context (e.g., 'Currently working on authentication refactor')."""

    OBSERVATION = "observation"
    """An observation or insight (e.g., 'User tends to ask follow-up questions')."""


class MemorySource(StrEnum):
    """How the memory was created."""

    EXPLICIT = "explicit"
    """User directly stated this fact."""

    INFERRED = "inferred"
    """Extracted/inferred from conversation."""

    SYSTEM = "system"
    """Created by system (consolidation, migration, etc.)."""


class Memory(BaseModel):
    """A memory item with full metadata."""

    id: UUID = Field(default_factory=uuid4)
    """Unique identifier for this memory."""

    text: str
    """The memory content (what we're remembering)."""

    durability: Durability = Durability.EPISODIC
    """How permanent this memory is."""

    confidence: float = Field(default=0.8, ge=0.0, le=1.0)
    """Confidence score (0-1) for inferred memories."""

    source: MemorySource = MemorySource.INFERRED
    """How this memory was created."""

    memory_type: MemoryType | None = None
    """Semantic type of this memory (fact, rule, decision, etc.)."""

    # Temporal validity
    valid_from: datetime = Field(default_factory=_utc_now)
    """When this fact became true."""

    valid_until: datetime | None = None
    """When this fact expires (None = permanent)."""

    # Version chain (for contradiction handling)
    supersedes: UUID | None = None
    """ID of the memory this one replaces (previous version)."""

    superseded_by: UUID | None = None
    """ID of the memory that replaced this one (next version)."""

    superseded_at: datetime | None = None
    """When this memory was superseded."""

    # Metadata
    created_at: datetime = Field(default_factory=_utc_now)
    """When this memory was stored."""

    last_accessed_at: datetime | None = None
    """Last time this memory was retrieved (for decay calculations)."""

    access_count: int = 0
    """Number of times this memory was retrieved."""

    tags: list[str] = Field(default_factory=list)
    """Optional tags for categorization."""

    metadata: dict[str, Any] = Field(default_factory=dict)
    """Additional metadata (source conversation, etc.)."""

    created_by: str | None = None
    """Who created this memory."""

    updated_by: str | None = None
    """Who last updated this memory."""

    def to_store_value(self) -> dict[str, Any]:
        """Convert to dict for storage in PostgresStore."""
        return {
            "id": str(self.id),
            "text": self.text,
            "durability": self.durability.value,
            "confidence": self.confidence,
            "source": self.source.value,
            "memory_type": self.memory_type.value if self.memory_type else None,
            "valid_from": self.valid_from.isoformat() if self.valid_from else None,
            "valid_until": self.valid_until.isoformat() if self.valid_until else None,
            "supersedes": str(self.supersedes) if self.supersedes else None,
            "superseded_by": str(self.superseded_by) if self.superseded_by else None,
            "superseded_at": self.superseded_at.isoformat() if self.superseded_at else None,
            "created_at": self.created_at.isoformat(),
            "last_accessed_at": (
                self.last_accessed_at.isoformat() if self.last_accessed_at else None
            ),
            "access_count": self.access_count,
            "tags": self.tags,
            "metadata": self.metadata,
            "created_by": self.created_by,
            "updated_by": self.updated_by,
        }

    @classmethod
    def from_store_value(cls, value: dict[str, Any]) -> "Memory":
        """Reconstruct from stored dict."""

        def parse_dt(v: str | None) -> datetime | None:
            return datetime.fromisoformat(v) if v else None

        def parse_uuid(v: str | None) -> UUID | None:
            return UUID(v) if v else None

        def parse_memory_type(v: str | None) -> MemoryType | None:
            return MemoryType(v) if v else None

        return cls(
            id=UUID(value["id"]),
            text=value["text"],
            durability=Durability(value["durability"]),
            confidence=value.get("confidence", 0.8),
            source=MemorySource(value.get("source", "inferred")),
            memory_type=parse_memory_type(value.get("memory_type")),
            valid_from=parse_dt(value.get("valid_from")) or _utc_now(),
            valid_until=parse_dt(value.get("valid_until")),
            supersedes=parse_uuid(value.get("supersedes")),
            superseded_by=parse_uuid(value.get("superseded_by")),
            superseded_at=parse_dt(value.get("superseded_at")),
            created_at=parse_dt(value.get("created_at")) or _utc_now(),
            last_accessed_at=parse_dt(value.get("last_accessed_at")),
            access_count=value.get("access_count", 0),
            tags=value.get("tags", []),
            metadata=value.get("metadata", {}),
            created_by=value.get("created_by"),
            updated_by=value.get("updated_by"),
        )

    def is_valid(self, at: datetime | None = None) -> bool:
        """Check if this memory is valid at the given time."""
        at = at or _utc_now()

        # Superseded memories are not valid
        if self.superseded_by is not None:
            return False

        # Check temporal validity
        if self.valid_from and at < self.valid_from:
            return False
        if self.valid_until and at > self.valid_until:
            return False

        return True

    def is_current(self) -> bool:
        """Check if this is the current version (not superseded)."""
        return self.superseded_by is None


class MemoryCreate(BaseModel):
    """Input for creating a new memory."""

    text: str
    durability: Durability = Durability.EPISODIC
    confidence: float = Field(default=0.8, ge=0.0, le=1.0)
    source: MemorySource = MemorySource.INFERRED
    memory_type: MemoryType | None = None
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None

    def to_memory(self) -> Memory:
        """Convert to full Memory object."""
        return Memory(
            text=self.text,
            durability=self.durability,
            confidence=self.confidence,
            source=self.source,
            memory_type=self.memory_type,
            valid_from=self.valid_from or _utc_now(),
            valid_until=self.valid_until,
            tags=self.tags,
            metadata=self.metadata,
            created_by=self.created_by,
        )


class MemoryUpdate(BaseModel):
    """Input for updating a memory (creates new version in chain)."""

    text: str
    confidence: float = Field(default=0.9, ge=0.0, le=1.0)
    source: MemorySource = MemorySource.EXPLICIT
    valid_from: datetime | None = None
    valid_until: datetime | None = None


class MemoryQuery(BaseModel):
    """Query parameters for memory retrieval."""

    query: str
    """Semantic search query."""

    limit: int = Field(default=10, ge=1, le=100)
    """Maximum number of results."""

    min_confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    """Minimum confidence threshold."""

    durability: list[Durability] | None = None
    """Filter by durability tier(s)."""

    memory_type: list[MemoryType] | None = None
    """Filter by memory type(s) (fact, rule, decision, etc.)."""

    include_superseded: bool = False
    """Include superseded (old version) memories."""

    include_expired: bool = False
    """Include memories past their valid_until date."""

    tags: list[str] | None = None
    """Filter by tags (any match)."""

    metadata_filter: dict[str, Any] | None = None
    """Filter by metadata key-value pairs (all must match)."""

    valid_at: datetime | None = None
    """Check validity at this time (default: now)."""


class MemoryPatch(BaseModel):
    """Patch metadata/tags/confidence without creating a version chain entry.

    Only non-None fields are applied. Use this for lightweight updates that
    don't warrant a new version in the chain (e.g., fixing tags, adjusting
    confidence, adding metadata).
    """

    tags: list[str] | None = None
    """Replace tags (None = no change)."""

    metadata: dict[str, Any] | None = None
    """Merge into existing metadata (None = no change)."""

    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    """Update confidence score (None = no change)."""

    durability: Durability | None = None
    """Update durability tier (None = no change)."""

    memory_type: MemoryType | None = None
    """Update memory type (None = no change)."""

    updated_by: str | None = None
    """Who is making this patch."""
